File: examples/hardware/tang_nano_20k/aurora_uart/client/aurora.py
#!/usr/bin/env python3
"""
aurora.py - Aurora UART control client for Tang Nano 20K.

Connects to the BL616 UART bridge on /dev/ttyUSB1 (or auto-detected when the
expected two-port Tang Nano bridge is the only ttyUSB device pair present) at
115200 8N1 and sends single-byte commands to the FPGA.

Requirements:
    pip install pyserial

Usage:
    python3 aurora.py [/dev/ttyUSB1]

Commands:
    a - aurora mode: smooth rainbow wave (default)
    r - red solid
    g - green solid
    b - blue solid
    f - faster (8x speed, same as holding S2 button)
    s - slower (back to 1x speed)
    x - freeze (hold current colours)
    q - quit

ACK: the FPGA echoes 'K' after each recognised command.
"""

import sys
import time

try:
    import serial
except ImportError:
    print("ERROR: pyserial not installed.  Run: pip install pyserial")
    sys.exit(1)

def _auto_detect_port() -> str:
    """Return the higher of the two Tang Nano ttyUSB ports when unambiguous."""
    import glob as _glob
    ports = sorted(_glob.glob("/dev/ttyUSB*"), key=lambda port: int(port.removeprefix("/dev/ttyUSB")))
    if not ports:
        print("ERROR: no /dev/ttyUSB* found — check USB cable and board power.")
        sys.exit(1)
    if len(ports) != 2:
        print(f"ERROR: refusing to auto-select a UART port from {len(ports)} ttyUSB devices.")
        print("  Pass the intended UART device explicitly, for example: python3 aurora.py /dev/ttyUSB1")
        print("  Available ports:")
        for port in ports:
            print(f"    {port}")
        sys.exit(1)
    return ports[1]

PORT = sys.argv[1] if len(sys.argv) > 1 else _auto_detect_port()
BAUD = 115200
VALID_CMDS = set("argbfsx")


def open_port(port: str) -> serial.Serial:
    try:
        ser = serial.Serial(port, BAUD, timeout=0.1)
        return ser
    except serial.SerialException as e:
        print(f"ERROR: cannot open {port}: {e}")
        print("  Check group access for the device.")
        print("  Prefer: sudo usermod -aG dialout $USER  (then log out and back in)")
        sys.exit(1)


def send_cmd(ser: serial.Serial, cmd: str) -> None:
    ser.write(bytes([ord(cmd)]))
    ser.flush()
    time.sleep(0.05)
    waiting = ser.in_waiting
    if waiting:
        ack = ser.read(waiting)
        ack_text = ack.decode("ascii", errors="replace").strip()
        print(f"  ACK: {ack_text!r}  ({ack.hex()})")
    else:
        print("  (no ACK received — check FPGA is running and /dev/ttyUSB1 is connected)")


def main() -> None:
    ser = open_port(PORT)
    print(f"Connected to {PORT} at {BAUD} baud.")
    print("Commands: a=aurora  r=red  g=green  b=blue  f=faster  s=slower  x=freeze  q=quit")
    print()

    while True:
        try:
            raw = input("> ").strip().lower()
        except (EOFError, KeyboardInterrupt):
            print()
            break

        if not raw:
            continue

        cmd = raw[0]
        if cmd == "q":
            break
        if cmd in VALID_CMDS:
            send_cmd(ser, cmd)
        else:
            print(f"  Unknown command '{cmd}'.  Valid: a r g b f s x q")

    ser.close()
    print("Disconnected.")


if __name__ == "__main__":
    main()

File: examples/hardware/tang_nano_20k/aurora_uart/client/aurora.ts
/**
 * aurora.ts - Aurora UART interactive client (Bun).
 *
 * Usage:  bun aurora.ts [port]       default port: auto-detected (highest /dev/ttyUSB*)
 *
 * Commands:
 *   a  aurora  - smooth rainbow wave (default)
 *   r  red     - solid red
 *   g  green   - solid green
 *   b  blue    - solid blue
 *   f  faster  - 8x speed
 *   s  slower  - 1x speed
 *   x  freeze  - hold current colours
 *   q  quit
 *
 * FPGA echoes 'K' for every recognised command.
 *
 * Compile hw:
 *   bun run apps/cli/src/index.ts compile \
 *     examples/hardware/tang_nano_20k/aurora_uart/hw \
 *     --board boards/tang_nano_20k.board.json --out .artifacts/aurora_uart --flash
 */

import * as readline from "readline";
import { openSync, writeSync, readSync, closeSync, constants, readdirSync } from "fs";
import { spawnSync } from "child_process";

const O_NOCTTY = 0o400;  // prevent port from becoming controlling terminal
const NB_FLAG = 0o4000; // O_NONBLOCK on Linux
const BAUD = 115_200;

function listTtyUsbPorts(): string[] {
  return readdirSync("/dev")
    .filter((entry) => /^ttyUSB\d+$/.test(entry))
    .sort((left, right) => Number(left.slice(6)) - Number(right.slice(6)))
    .map((entry) => `/dev/${entry}`);
}

// Auto-detect only when the expected two-port Tang Nano bridge is present.
function resolvePort(arg: string | undefined): string {
  if (arg) return arg;
  const ports = listTtyUsbPorts();
  if (ports.length === 0) {
    console.error("No /dev/ttyUSB* found — check USB cable and board power.");
    process.exit(1);
  }
  if (ports.length !== 2) {
    console.error(`Refusing to auto-select a UART port from ${ports.length} ttyUSB devices.`);
    console.error("Pass the intended UART device explicitly, for example: bun aurora.ts /dev/ttyUSB1");
    console.error(`Available ports: ${ports.join(", ")}`);
    process.exit(1);
  }
  return ports[1];
}
const PORT = resolvePort(process.argv[2]);

const CMDS: Record<string, { byte: string; label: string }> = {
  a: { byte: "a", label: "aurora (rainbow wave)" },
  aurora: { byte: "a", label: "aurora (rainbow wave)" },
  r: { byte: "r", label: "red solid" },
  red: { byte: "r", label: "red solid" },
  g: { byte: "g", label: "green solid" },
  green: { byte: "g", label: "green solid" },
  b: { byte: "b", label: "blue solid" },
  blue: { byte: "b", label: "blue solid" },
  f: { byte: "f", label: "faster (8x)" },
  faster: { byte: "f", label: "faster (8x)" },
  s: { byte: "s", label: "slower (1x)" },
  slower: { byte: "s", label: "slower (1x)" },
  x: { byte: "x", label: "freeze" },
  freeze: { byte: "x", label: "freeze" },
};

// IMPORTANT: open FIRST, then stty.
// USB-serial drivers (BL616, FTDI) reinitialize baud rate on open.
// Running stty before openSync means the driver resets to 9600 on open.
let fd: number;
try {
  fd = openSync(PORT, constants.O_RDWR | O_NOCTTY | NB_FLAG);
} catch (e: any) {
  console.error(`Cannot open ${PORT}: ${e.message}`);
  console.error("  Check group access for the device and prefer adding your user to the dialout group.");
  process.exit(1);
}

const stty = spawnSync("stty", [
  "-F", PORT, `${BAUD}`, "raw", "cs8", "-cstopb", "-parenb", "clocal", "cread", "-echo", "-crtscts",
]);
if (stty.status !== 0) {
  console.error(`Cannot configure ${PORT}: ${stty.stderr?.toString().trim()}`);
  process.exit(1);
}

console.log(`Aurora UART  |  ${PORT} @ ${BAUD} baud`);
console.log("Commands: a=aurora  r=red  g=green  b=blue  f=faster  s=slower  x=freeze  q=quit");
console.log();

// Poll for FPGA ACK bytes every 20 ms without blocking readline.
const rxBuf = Buffer.alloc(16);
function poll() {
  try {
    const n = readSync(fd, rxBuf, 0, rxBuf.length, null);
    if (n > 0) {
      const s = rxBuf.subarray(0, n).toString("ascii").replace(/[^\x20-\x7e]/g, "");
      if (s) process.stdout.write(`\r  << ACK '${s}'\n> `);
    }
  } catch (e: any) {
    if (e.code !== "EAGAIN" && e.code !== "EWOULDBLOCK") {
      console.error(`\nRead error: ${e.message}`);
    }
  }
  setTimeout(poll, 20);
}
poll();

const rl = readline.createInterface({
  input: process.stdin, output: process.stdout,
  prompt: "> ", terminal: true,
});
rl.prompt();

rl.on("line", (line) => {
  const cmd = line.trim().toLowerCase();
  if (!cmd) { rl.prompt(); return; }
  if (cmd === "q" || cmd === "quit") { closeSync(fd); process.exit(0); }
  const entry = CMDS[cmd];
  if (entry) {
    console.log(`  >> ${entry.label}`);
    writeSync(fd, Buffer.from([entry.byte.charCodeAt(0)]));
  } else {
    console.log(`  Unknown: '${cmd}'.  Valid: a r g b f s x q`);
  }
  rl.prompt();
});

rl.on("close", () => { closeSync(fd); process.exit(0); });
